<center><h3>VELLORE INSTITUTE OF TECHNOLOGY</h3></center>
<center><b>CSE4020	Machine Learning<b></center>
<center><b>Lab Assessment - 2<b></center>

<h1>17BCE0581</h1>
<h3>SATYAM SINGH CHAUHAN</h3>

<h1>Multiple Linear Regression</h1>
<h2>Importing the Required Libraries</h2>
<ul>
    <li>matplotlib.pyplot</li>
    <li>pandas</li>
    <li>pylab</li>
    <li>numpy</li>
</ul>

In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import pylab as pl
import numpy as np
%matplotlib inline
from sklearn import datasets

<b>Load the diabetes dataset</b>


In [2]:
diabetes = datasets.load_diabetes()

In [3]:
data1 = pd.DataFrame(data = np.c_[diabetes['data'],diabetes['target']],columns = diabetes['feature_names'] + ['target'])

<b>Emploring and Studying the Datasets using </b><i>head</i> <b>and</b> <i>describe</i> <b>functions</b>

In [4]:
data1.head(10)

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
5 -0.092695 -0.044642 -0.040696 -0.019442 -0.068991 -0.079288  0.041277   
6 -0.045472  0.050680 -0.047163 -0.015999 -0.040096 -0.024800  0.000779   
7  0.063504  0.050680 -0.001895  0.066630  0.090620  0.108914  0.022869   
8  0.041708  0.050680  0.061696 -0.040099 -0.013953  0.006202 -0.028674   
9 -0.070900 -0.044642  0.039062 -0.033214 -0.012577 -0.034508 -0.024993   

         s4        s5        s6  target  
0 -0.002592  0.019908 -0.017646   151.0  
1 -0.039493 -0.068330 -0.092204    75.0  
2 -0.002592  0.002864 -0.025930   141.0  
3  0.034309  0.022692 -0.009362   206.0  
4 -0.002592 -0.031991 -0.046641   135.0  
5 -0.076395 -0.041180 -0.096346    97.0  
6 -0.039493 -0.062913 -0.038357   138.0  
7  0.017703 -0.035817  0.003064    63.0  
8 -0.002592 -0.014956  0.011349   110.0  
9 -0.002592  0.067736 -0.013504   310.0

In [5]:
data1.describe()

age           sex           bmi            bp            s1  \
count  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02   
mean  -3.634285e-16  1.308343e-16 -8.045349e-16  1.281655e-16 -8.835316e-17   
std    4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02   
min   -1.072256e-01 -4.464164e-02 -9.027530e-02 -1.123996e-01 -1.267807e-01   
25%   -3.729927e-02 -4.464164e-02 -3.422907e-02 -3.665645e-02 -3.424784e-02   
50%    5.383060e-03 -4.464164e-02 -7.283766e-03 -5.670611e-03 -4.320866e-03   
75%    3.807591e-02  5.068012e-02  3.124802e-02  3.564384e-02  2.835801e-02   
max    1.107267e-01  5.068012e-02  1.705552e-01  1.320442e-01  1.539137e-01   

                 s2            s3            s4            s5            s6  \
count  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02   
mean   1.327024e-16 -4.574646e-16  3.777301e-16 -3.830854e-16 -3.412882e-16   
std    4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02   
min   -1.156131e-01 -1.023071e-01 -7.639450e-02 -1.260974e-01 -1.377672e-01   
25%   -3.035840e-02 -3.511716e-02 -3.949338e-02 -3.324879e-02 -3.317903e-02   
50%   -3.819065e-03 -6.584468e-03 -2.592262e-03 -1.947634e-03 -1.077698e-03   
75%    2.984439e-02  2.931150e-02  3.430886e-02  3.243323e-02  2.791705e-02   
max    1.987880e-01  1.811791e-01  1.852344e-01  1.335990e-01  1.356118e-01   

           target  
count  442.000000  
mean   152.133484  
std     77.093005  
min     25.000000  
25%     87.000000  
50%    140.500000  
75%    211.500000  
max    346.000000

<b>Filtering the required Data</b>

In [78]:
data2 = data1[['bmi','age','bp','s1','s5','s6','target']]
data3 = data1[['bmi','age','sex','bp','s1','s3','s5','s6','target']]
data4 = data1[['bmi','age','bp','s1','s3','s5','target']]

<b>Exploring and Studying the Data2</b>

In [56]:
data2.head()

bmi       age        bp        s1        s5        s6  target
0  0.061696  0.038076  0.021872 -0.044223  0.019908 -0.017646   151.0
1 -0.051474 -0.001882 -0.026328 -0.008449 -0.068330 -0.092204    75.0
2  0.044451  0.085299 -0.005671 -0.045599  0.002864 -0.025930   141.0
3 -0.011595 -0.089063 -0.036656  0.012191  0.022692 -0.009362   206.0
4 -0.036385  0.005383  0.021872  0.003935 -0.031991 -0.046641   135.0

<b>Exploring and Studying the Data3</b>

In [79]:
data3.head()

bmi       age       sex        bp        s1        s3        s5  \
0  0.061696  0.038076  0.050680  0.021872 -0.044223 -0.043401  0.019908   
1 -0.051474 -0.001882 -0.044642 -0.026328 -0.008449  0.074412 -0.068330   
2  0.044451  0.085299  0.050680 -0.005671 -0.045599 -0.032356  0.002864   
3 -0.011595 -0.089063 -0.044642 -0.036656  0.012191 -0.036038  0.022692   
4 -0.036385  0.005383 -0.044642  0.021872  0.003935  0.008142 -0.031991   

         s6  target  
0 -0.017646   151.0  
1 -0.092204    75.0  
2 -0.025930   141.0  
3 -0.009362   206.0  
4 -0.046641   135.0

<b>Exploring and Studying the Data4</b>

In [50]:
data4.head()

bmi       age        bp        s1        s3        s5  target
0  0.061696  0.038076  0.021872 -0.044223 -0.043401  0.019908   151.0
1 -0.051474 -0.001882 -0.026328 -0.008449  0.074412 -0.068330    75.0
2  0.044451  0.085299 -0.005671 -0.045599 -0.032356  0.002864   141.0
3 -0.011595 -0.089063 -0.036656  0.012191 -0.036038  0.022692   206.0
4 -0.036385  0.005383  0.021872  0.003935  0.008142 -0.031991   135.0

<b>Split the data into training/testing sets
</b>

In [62]:
split2 = np.random.rand(len(data2)) < 0.8
train_data2 = data2[split2]
test_data2 = data2[~split2]

In [89]:
split3 = np.random.rand(len(data3)) < 0.9
train_data3 = data3[split3]
test_data3 = data3[~split3]

In [51]:
split4 = np.random.rand(len(data4)) < 0.8
train_data4 = data4[split4]
test_data4 = data4[~split4]

<b>Create linear regression object<b>
    <br>
<b>Train the model using the training sets</b>
    <br>
<b>And Display The coefficients</b>


In [63]:
from sklearn import linear_model
slr2 = linear_model.LinearRegression()
train_x2 = np.asanyarray(train_data2[['bmi','age','bp','s1','s5','s6']])
train_y2 = np.asanyarray(train_data2[["target"]])
slr2.fit(train_x2,train_y2)
print('Coefficient: ', slr2.coef_)
print('Intercept: ',slr2.intercept_)

Coefficient:  [[ 583.39946551 -102.30526358  333.3440783  -204.2936595   616.64781099
    30.90767484]]
Intercept:  [152.97079742]


In [90]:
from sklearn import linear_model
slr3 = linear_model.LinearRegression()
train_x3 = np.asanyarray(train_data3[['bmi','age','sex','bp','s1','s3','s5','s6']])
train_y3 = np.asanyarray(train_data3[["target"]])
slr3.fit(train_x3,train_y3)
print('Coefficient: ', slr3.coef_)
print('Intercept: ',slr3.intercept_)

Coefficient:  [[ 526.11190048    9.96222165 -233.96335056  319.7649089  -158.65303983
  -207.39184311  521.64182454  122.6196994 ]]
Intercept:  [150.78480642]


In [52]:
from sklearn import linear_model
slr4 = linear_model.LinearRegression()
train_x4 = np.asanyarray(train_data4[['bmi','age','bp','s1','s3','s5']])
train_y4 = np.asanyarray(train_data4[["target"]])
slr4.fit(train_x4,train_y4)
print('Coefficient: ', slr4.coef_)
print('Intercept: ',slr4.intercept_)

Coefficient:  [[ 481.00809091  -39.47417345  350.17229208 -148.41499095 -136.16664934
   573.86448337]]
Intercept:  [152.03787327]


Model 1
<b>Make predictions using the testing set</b>
# Data 2

In [64]:
from sklearn.metrics import r2_score, mean_squared_error
test_x2 = np.asanyarray(test_data2[['bmi','age','bp','s1','s5','s6']])
test_y2 = np.asanyarray(test_data2[["target"]])
test_result2 = slr2.predict(test_x2)

<b>Calculating R2 Score for Data 2</b>

In [65]:
r2Score_2 = r2_score(test_y2, test_result2)
print('R2 Score: ',r2Score_2)

R2 Score:  0.49244783525800795


Model 2
<b>Make predictions using the testing set</b>
# Data 3

In [91]:
from sklearn.metrics import r2_score, mean_squared_error
test_x3 = np.asanyarray(test_data3[['bmi','age','sex','bp','s1','s3','s5','s6']])
test_y3 = np.asanyarray(test_data3[["target"]])
test_result3 = slr3.predict(test_x3)

<b>Calculating R2 Score for Data 3</b>

In [92]:
r2Score_3 = r2_score(test_y3, test_result3)
print('R2 Score: ',r2Score_3)

R2 Score:  0.5583133109733476


Model 3
<b>Make predictions using the testing set</b>
# Data 4

In [53]:
from sklearn.metrics import r2_score, mean_squared_error
test_x4 = np.asanyarray(test_data4[['bmi','age','bp','s1','s3','s5']])
test_y4 = np.asanyarray(test_data4[["target"]])
test_result4 = slr4.predict(test_x4)

<b>Calculating R2 Score for Data 4</b>

In [54]:
r2Score_4 = r2_score(test_y4, test_result4)
print('R2 Score: ',r2Score_4)

R2 Score:  0.5089621667814918


<hr>

Best R2 Score Observed : <b>0.5583133109733476</b>
of Data 3

<h3>Model 2 when compared, found out to be the best was with the Data 3 and attributes ('bmi','age','sex','bp','s1','s3','s5','s6')</h3>